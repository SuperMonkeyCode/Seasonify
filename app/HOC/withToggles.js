/* @flow */
import { Button } from 'native-base';
import { compose, getDisplayName } from 'recompose';
import { connect } from 'react-redux';
import React from 'react';

type Toggle = {
  stateKey: string,
  selector: Function,
  action: Function,
  buttonFacets: { [facet: string | boolean]: React$Element<*> },
};

type Props = {
  /* HOC param */
  toggles: Toggle[],
  /* HOC */
  toggleButtons?: [Button],
  /* connect */
  selectorX: SortingValue,
  actionX: Function,
};

export const withToggles = (toggles: Toggle[]) => (
  BaseComponent: ReactClass<*>,
): ReactClass<*> =>
  class ComponentWithToggles extends React.Component<void, Props, void> {
    static displayName = `withToggles(${getDisplayName(BaseComponent)})`;

    toggles: Toggle[];
    togglesAction: Array<Function>;

    constructor(props) {
      super(props);

      this.toggles = toggles;

      this.togglesAction = this.toggles.map((toggle, index) => () =>
        this.props[`action${index}`](toggle.stateKey),
      );
    }

    render() {
      return (
        <BaseComponent
          {...this.props}
          toggleButtons={this.toggles.map((toggle, index) =>
            (<Button
              key={`${toggle.selector.name}_${toggle.stateKey}`}
              transparent
              small
              onPress={this.togglesAction[index]}
            >
              {toggle.buttonFacets[this.props[`selector${index}`]]}
            </Button>),
          )}
        />
      );
    }
  };

export const makeMapStateToProps = (state: Object, toggles: Toggle[]) =>
  toggles.reduce(
    (selectors, toggle, index) => ({
      ...selectors,
      [`selector${index}`]: toggle.selector(state, toggle.stateKey),
    }),
    {},
  );

export const makeMapActionsToProps = (toggles: Toggle[]) =>
  toggles.reduce(
    (actions, toggle, index) => ({ ...actions, [`action${index}`]: toggle.action }),
    {},
  );

const redux = (toggles: Toggle[]) =>
  connect(state => makeMapStateToProps(state, toggles), makeMapActionsToProps(toggles));

export default (toggles: Toggle[]) => compose(redux(toggles), withToggles(toggles));
