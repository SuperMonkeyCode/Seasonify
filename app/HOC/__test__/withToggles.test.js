import { Text, View } from 'react-native';
import React from 'react';

import { _withToggles, makeMapStateToProps, makeMapActionsToProps } from '../withToggles';

const Component = ({ title, toggleButtons }) =>
  (<View>
    <Text>
      {title}
    </Text>
    {toggleButtons}
  </View>);

describe('With a 2 toggles and passing some title', () => {
  const toggles = [
    {
      stateKey: 'statekey0',
      selector: () => 'ASC',
      action: jest.fn(),
      buttonFacets: { ASC: <Text>A-Z</Text>, DESC: <Text>Z-A</Text> },
    },
    {
      stateKey: 'statekey1',
      selector: () => 'DESC',
      action: jest.fn(),
      buttonFacets: { ASC: <Text>up</Text>, DESC: <Text>down</Text> },
    },
  ];
  const ComponentWithToggles = _withToggles(toggles)(Component);

  const connectProps = {
    selector0: 'ASC',
    selector1: 'DESC',
    action0: jest.fn(),
    action1: jest.fn(),
  };

  const Hoc = shallow(<ComponentWithToggles title="Some title" {...connectProps} />);

  it('should match the rendered children', () => {
    expect(Hoc).toMatchSnapshot();
  });

  it('should have the right component display name', () => {
    expect(ComponentWithToggles.displayName).toEqual('withToggles(Component)');
  });

  it('gives children the title prop', () => {
    expect(Hoc.props().title).toEqual('Some title');
  });

  it('gives children the toggle button', () => {
    expect(Hoc.props().toggleButtons.length).toEqual(2);
  });

  it('calls associated action when pressing the first button', () => {
    Hoc.props().toggleButtons[0].props.onPress();
    expect(connectProps.action0).toBeCalled();
  });
});

describe('Connect() forging with 2 toggles', () => {
  const toggles = [
    {
      stateKey: 'statekey0',
      selector: jest.fn().mockReturnValue('ASC'),
      action: jest.fn(),
    },
    {
      stateKey: 'statekey1',
      selector: jest.fn().mockReturnValue('DESC'),
      action: jest.fn(),
    },
  ];
  const mapStateToProps = makeMapStateToProps({}, toggles);
  const mapActionsToProps = makeMapActionsToProps(toggles);

  it('maps state to props', () => {
    expect(mapStateToProps).toEqual({
      selector0: 'ASC',
      selector1: 'DESC',
    });
    expect(toggles[0].selector).toBeCalledWith({}, toggles[0].stateKey);
    expect(toggles[1].selector).toBeCalledWith({}, toggles[1].stateKey);
  });

  it('maps actions to props', () => {
    expect(mapActionsToProps).toEqual({
      action0: toggles[0].action,
      action1: toggles[1].action,
    });
  });
});
