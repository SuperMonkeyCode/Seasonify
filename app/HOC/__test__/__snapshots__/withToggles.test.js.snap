// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`With a 2 toggles and passing some title should match the rendered children 1`] = `
<Component
  action0={[Function]}
  action1={[Function]}
  selector0="ASC"
  selector1="DESC"
  title="Some title"
  toggleButtons={
    Array [
      <Styled(Button)
        onPress={[Function]}
        small={true}
        transparent={true}
    >
        <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
        >
            A-Z
        </Text>
    </Styled(Button)>,
      <Styled(Button)
        onPress={[Function]}
        small={true}
        transparent={true}
    >
        <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
        >
            down
        </Text>
    </Styled(Button)>,
    ]
  }
/>
`;
